nip.py

```diff
@@ -68,35 +69,47 @@ def var_deps(expr: source.ExprT[source.ProgVarName]) -> source.ExprT[GuardVarNam
     return reduce(source.expr_and, map(guard_var, source.all_vars_in_expr(expr)), source.expr_true)
 
 
-def make_state_update_for_node(node: source.Node[source.ProgVarName]) -> Iterator[source.Update[GuardVarName]]:
+def make_state_update_for_node(node: source.Node[source.ProgVarName], new_variables: Set[source.ExprVarT[GuardVarName]]) -> Iterator[source.Update[GuardVarName]]:
     if isinstance(node, source.NodeBasic):
         for upd in node.upds:
             if not source.is_loop_counter_name(upd.var.name):
+                # new variables go in LHS
+                new_variables.add(guard_var(upd.var))
                 yield source.Update(guard_var(upd.var), var_deps(upd.expr))
     elif isinstance(node, source.NodeCall):
         deps = reduce(source.expr_and, (var_deps(arg)
                                         for arg in node.args), source.expr_true)
         for ret in node.rets:
             assert not source.is_loop_counter_name(
                 ret.name), "didn't expect a return value to be a loop counter"
+            new_variables.add(guard_var(ret))  # new variables go in LHS
             yield source.Update(guard_var(ret), deps)
     else:
         assert not isinstance(node, source.NodeEmpty | source.NodeCond |
                               source.NodeAssume | source.NodeAssert), "doesn't make sense to have a state update for those nodes"
         assert_never(node)
 
 
-def make_protection_for_node(node: source.Node[source.ProgVarName]) -> source.ExprT[GuardVarName]:
+def make_protection_for_node(node: source.Node[source.ProgVarName]) -> Tuple[Set[source.ExprVarT[GuardVarName]], source.ExprT[GuardVarName]]:
+    variables: Set[source.ExprVarT[GuardVarName]] = set([])
+    guards: Tuple[source.ExprVarT[GuardVarName], ...] = tuple([guard_var(
```

Comment:
`tuple(f(x) for x in xs)` is better than `tuple([f(x) for x in xs])` because the later one allocates the entire list, and then converts it to a tuple. The first one creates an iterator (consumes O(1) memory) to generate each element, and turns it into a tuple directly.